### PR TITLE
[SPARK-24983][Catalyst] Add configuration for maximum number of leaf expressions in collapsed projects

### PR DESCRIPTION
**Pull request by dvogelbacher**
## What changes were proposed in this pull request?

Add a configuration option for the maximum number of leaf expressions in collapsed project nodes. If a collapsed project node (result of the `CollapseProject` optimizer rule) would have more leaf expressions than the configured maximum number we don't collapse. This is to protect against an exponentially exploding number of leaf expressions when collapsing many (binary) expression that refer to the same columns (see https://issues.apache.org/jira/browse/SPARK-24983).

## How was this patch tested?
Add a new unit test.

### CODE REVIEW

**Comment by dvogelbacher** [comment]
@HyukjinKwon can you help with finding reviewers for this PR?

**Comment by HyukjinKwon** [comment]
Usually @gatorsmile and @cloud-fan.

**Comment by gatorsmile** [comment]
Let us blacklist CASE WHEN in CollapseProject, instead of introducing this new conf.

**Comment by gatorsmile** [comment]
@dvogelbacher Currently, in the master branch (2.4 release), you have a workaround. Add CollapseProject to `spark.sql.optimizer.excludedRules` before such queries.

**Comment by dvogelbacher** [comment]
@gatorsmile yes, I found that workaround. Very useful :)
I think it would still be good to handle this better by default. I can see that introducing such an arbitrary configuration param doesn't seem optimal and am open to better suggestions.

Not sure if blacklisting case-when statements outright is the right way to go. That could have negative perf impacts as well? And it wouldn't handle the case in the unit test where we have exponential growth when adding/subtracting columns (though that example might be somewhat contrived). 

Maybe we should just not collapse if the number of leaf expressions in the collapsed project is higher than the sum of the number of leaf expressions in the original statements?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -631,19 +631,26 @@ object ColumnPruning extends Rule[LogicalPlan] {
 object CollapseProject extends Rule[LogicalPlan] {
 
   def apply(plan: LogicalPlan): LogicalPlan = plan transformUp {
-    case p1 @ Project(_, p2: Project) =>
-      if (haveCommonNonDeterministicOutput(p1.projectList, p2.projectList)) {
-        p1
-      } else {
-        p2.copy(projectList = buildCleanedProjectList(p1.projectList, p2.projectList))
-      }
+    case p1@Project(_, p2: Project) =>
```
Comment (kiszk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: nit: Do we need to change this line? We can keep this line as is.

**Comment by gatorsmile** [comment]
Backlisting case-when statements looks good to me.

**Comment by HyukjinKwon** [comment]
Closing due to inactivity over an year.

**Pull request closed without merging**